File: frontend/src/types/pipeline.ts
/**
 * TypeScript interfaces mirroring the backend Pydantic models.
 * These define the contract between frontend and backend.
 */

// ---------------------------------------------------------------------------
// Enums
// ---------------------------------------------------------------------------

export type StageType = 'extract' | 'transform' | 'load';
export type ExtractSource = 'csv' | 'json' | 'parquet';
export type LoadSink = 'csv' | 'json' | 'parquet';
export type TransformOperation =
  | 'filter'
  | 'select'
  | 'rename'
  | 'group_by'
  | 'sort'
  | 'cast'
  | 'join'
  | 'derive'
  | 'drop_nulls'
  | 'unique';

export type FilterOperator =
  | 'eq' | 'ne' | 'gt' | 'gte' | 'lt' | 'lte'
  | 'in' | 'not_in' | 'is_null' | 'is_not_null'
  | 'contains' | 'starts_with' | 'ends_with';

export type AggFunction = 'sum' | 'mean' | 'count' | 'min' | 'max' | 'first' | 'last';
export type JoinHow = 'inner' | 'left' | 'right' | 'outer' | 'cross';

// ---------------------------------------------------------------------------
// Config models
// ---------------------------------------------------------------------------

export interface ExtractConfig {
  path: string;
  separator?: string;
  has_header?: boolean;
  dtypes?: Record<string, string>;
  n_rows?: number;
}

export interface FilterCondition {
  field: string;
  operator: FilterOperator;
  value?: unknown;
}

export interface AggregationSpec {
  column: string;
  function: AggFunction;
  alias?: string;
}

export interface LoadConfig {
  path: string;
  separator?: string;
  compression?: 'snappy' | 'gzip' | 'lz4' | 'zstd' | 'none';
}

// ---------------------------------------------------------------------------
// Stage
// ---------------------------------------------------------------------------

export interface StageConfig {
  id: string;
  type: StageType;
  depends_on?: string[];

  // Extract
  source?: ExtractSource;
  extract_config?: ExtractConfig;

  // Transform
  operation?: TransformOperation;
  transform_config?: Record<string, unknown>;

  // Load
  sink?: LoadSink;
  load_config?: LoadConfig;

  // UI metadata
  position?: { x: number; y: number };
  label?: string;
}

export interface SelectConfig {
  columns: string[];
}

export interface RenameConfig {
  mapping: Record<string, string>;
}

export interface GroupByConfig {
  group_by: string[];
  aggregations: AggregationSpec[];
}

export interface SortConfig {
  by: string[];
  descending?: boolean | boolean[];
}

export interface CastConfig {
  mapping: Record<string, string>;
}

export interface JoinConfig {
  right_stage: string;
  on?: string[];
  left_on?: string[];
  right_on?: string[];
  how?: JoinHow;
}

export interface DeriveConfig {
  columns: Record<string, string>;
}

export interface DropNullsConfig {
  subset?: string[];
}

export interface UniqueConfig {
  subset?: string[];
  keep?: 'first' | 'last' | 'any' | 'none';
}

export interface FilterConfig {
  conditions: FilterCondition[];
  logic?: 'and' | 'or';
}

// ---------------------------------------------------------------------------
// Pipeline
// ---------------------------------------------------------------------------

export interface PipelineMetadata {
  name: string;
  description?: string;
  version?: string;
  variables?: Record<string, unknown>;
}

export interface PipelineConfig {
  pipeline: PipelineMetadata;
  stages: StageConfig[];
}

// ---------------------------------------------------------------------------
// API responses
// ---------------------------------------------------------------------------

export interface StageResult {
  stage_id: string;
  status: 'success' | 'error' | 'skipped';
  rows_affected: number;
  duration_ms: number;
  error?: string | null;
}

export interface PipelineRunResult {
  pipeline_id: string;
  pipeline_name: string;
  status: 'success' | 'error' | 'partial';
  total_duration_ms: number;
  stage_results: StageResult[];
  error?: string | null;
}

export interface PipelineListItem {
  id: string;
  name: string;
  description: string;
  stage_count: number;
  last_run_status?: string | null;
  last_run_at?: string | null;
  created_at: string;
  updated_at: string;
}

export interface StageTypeInfo {
  type: StageType;
  category: string;
  name: string;
  description: string;
  icon: string;
  config_schema: Record<string, unknown>;
}

export interface PipelineDetail {
  id: string;
  name: string;
  description: string;
  yaml_config: string;
  stage_count: number;
  created_at: string;
  updated_at: string;
}

export interface ValidationResult {
  valid: boolean;
  pipeline_name?: string;
  stage_count?: number;
  error?: string;
}

File: backend/app/models/pipeline.py
"""
Pydantic models for Poloom pipeline YAML configuration.

These models define the schema contract between the frontend editor
and the backend executor. They validate YAML configs at ingestion time.
"""

from __future__ import annotations

import enum
from typing import Any, Literal, Optional

from pydantic import BaseModel, Field


# ---------------------------------------------------------------------------
# Enums
# ---------------------------------------------------------------------------

class StageType(str, enum.Enum):
    EXTRACT = "extract"
    TRANSFORM = "transform"
    LOAD = "load"


class ExtractSource(str, enum.Enum):
    CSV = "csv"
    JSON = "json"
    PARQUET = "parquet"


class LoadSink(str, enum.Enum):
    CSV = "csv"
    JSON = "json"
    PARQUET = "parquet"


class TransformOperation(str, enum.Enum):
    FILTER = "filter"
    SELECT = "select"
    RENAME = "rename"
    GROUP_BY = "group_by"
    SORT = "sort"
    CAST = "cast"
    JOIN = "join"
    DERIVE = "derive"
    DROP_NULLS = "drop_nulls"
    UNIQUE = "unique"


class AggFunction(str, enum.Enum):
    SUM = "sum"
    MEAN = "mean"
    COUNT = "count"
    MIN = "min"
    MAX = "max"
    FIRST = "first"
    LAST = "last"


# ---------------------------------------------------------------------------
# Config sub-models
# ---------------------------------------------------------------------------

class AggregationSpec(BaseModel):
    """A single aggregation within a group_by stage."""
    column: str
    function: AggFunction
    alias: Optional[str] = None


class ExtractConfig(BaseModel):
    """Configuration for extract stages."""
    path: str
    separator: str = ","
    has_header: bool = True
    dtypes: Optional[dict[str, str]] = None
    n_rows: Optional[int] = None


class FilterCondition(BaseModel):
    """A single filter condition."""
    field: str
    operator: Literal["eq", "ne", "gt", "gte", "lt", "lte", "in", "not_in", "is_null", "is_not_null", "contains", "starts_with", "ends_with"]
    value: Any = None


class FilterConfig(BaseModel):
    """Configuration for filter transform."""
    conditions: list[FilterCondition]
    logic: Literal["and", "or"] = "and"


class SelectConfig(BaseModel):
    """Configuration for select transform."""
    columns: list[str]


class RenameConfig(BaseModel):
    """Configuration for rename transform."""
    mapping: dict[str, str]


class GroupByConfig(BaseModel):
    """Configuration for group_by transform."""
    group_by: list[str]
    aggregations: list[AggregationSpec]


class SortConfig(BaseModel):
    """Configuration for sort transform."""
    by: list[str]
    descending: list[bool] | bool = False


class CastConfig(BaseModel):
    """Configuration for cast transform."""
    mapping: dict[str, str]  # column -> target dtype


class JoinConfig(BaseModel):
    """Configuration for join transform."""
    right_stage: str  # id of the other stage to join with
    on: list[str] | None = None
    left_on: list[str] | None = None
    right_on: list[str] | None = None
    how: Literal["inner", "left", "right", "outer", "cross"] = "inner"


class DeriveConfig(BaseModel):
    """Configuration for derive (add column) transform."""
    columns: dict[str, str]  # alias -> expression string


class DropNullsConfig(BaseModel):
    """Configuration for drop_nulls transform."""
    subset: list[str] | None = None


class UniqueConfig(BaseModel):
    """Configuration for unique transform."""
    subset: list[str] | None = None
    keep: Literal["first", "last", "any", "none"] = "first"


class LoadConfig(BaseModel):
    """Configuration for load stages."""
    path: str
    separator: str = ","
    compression: Optional[Literal["snappy", "gzip", "lz4", "zstd", "none"]] = None


# ---------------------------------------------------------------------------
# Stage model
# ---------------------------------------------------------------------------

class StageConfig(BaseModel):
    """A single stage in the ETL pipeline."""
    id: str = Field(..., pattern=r"^[a-zA-Z_][a-zA-Z0-9_]*$")
    type: StageType
    depends_on: list[str] = Field(default_factory=list)

    # Extract fields
    source: Optional[ExtractSource] = None
    extract_config: Optional[ExtractConfig] = None

    # Transform fields
    operation: Optional[TransformOperation] = None
    transform_config: Optional[dict[str, Any]] = None

    # Load fields
    sink: Optional[LoadSink] = None
    load_config: Optional[LoadConfig] = None

    # UI metadata (positions for React Flow)
    position: Optional[dict[str, float]] = None
    label: Optional[str] = None


# ---------------------------------------------------------------------------
# Pipeline model
# ---------------------------------------------------------------------------

class PipelineMetadata(BaseModel):
    """Top-level pipeline metadata."""
    name: str
    description: str = ""
    version: str = "1.0"
    variables: dict[str, Any] = Field(default_factory=dict)


class PipelineConfig(BaseModel):
    """Complete pipeline configuration — maps directly to a YAML file."""
    pipeline: PipelineMetadata
    stages: list[StageConfig]


# ---------------------------------------------------------------------------
# API response models
# ---------------------------------------------------------------------------

class StageResult(BaseModel):
    """Result of executing a single stage."""
    stage_id: str
    status: Literal["success", "error", "skipped"]
    rows_affected: int = 0
    duration_ms: float = 0.0
    error: Optional[str] = None


class PipelineRunResult(BaseModel):
    """Result of a full pipeline execution."""
    pipeline_id: str
    pipeline_name: str
    status: Literal["success", "error", "partial"]
    total_duration_ms: float = 0.0
    stage_results: list[StageResult] = Field(default_factory=list)
    error: Optional[str] = None


class PipelineListItem(BaseModel):
    """Summary of a pipeline for listing."""
    id: str
    name: str
    description: str = ""
    stage_count: int = 0
    last_run_status: Optional[str] = None
    last_run_at: Optional[str] = None
    created_at: str = ""
    updated_at: str = ""


class StageTypeInfo(BaseModel):
    """Info about a stage type for the frontend palette."""
    type: StageType
    category: str
    name: str
    description: str
    icon: str
    config_schema: dict[str, Any] = Field(default_factory=dict)
